src/core/providers/vpc/qgsvirtualpointcloudprovider.cpp

```diff
@@ -0,0 +1,555 @@
+      si.extent = QgsRectangle( nativeBbox[0].get<double>(), nativeBbox[1].get<double>(),
+                                nativeBbox[3].get<double>(), nativeBbox[4].get<double>() );
+    }
+    else if ( f.contains( "bbox" ) && mCrs.isValid() )
+    {
+      nlohmann::json bboxWgs = f["bbox"];
+      QgsRectangle bbox = QgsRectangle( bboxWgs[0].get<double>(), bboxWgs[1].get<double>(),
+                                        bboxWgs[2].get<double>(), bboxWgs[3].get<double>() );
+
+      try
+      {
+        si.extent = transform.transformBoundingBox( bbox );
+      }
+      catch ( QgsCsException & )
+      {
+        QgsDebugMsg( QStringLiteral( "Cannot transform bbox to layer crs." ) );
+        continue;
+      }
+    }
+    else
+    {
+      QgsDebugMsg( QStringLiteral( "Missing extent information." ) );
+      continue;
+    }
+
+    const auto geom = f["geometry"];
+    try
+    {
+      QgsMultiPolygonXY geometry;
+      if ( geom.at( "type" ) == "Polygon" )
+      {
+        QgsPolygonXY polygon;
+        for ( auto &geomRing : geom.at( "coordinates" ) )
+        {
+          QgsPolylineXY ring;
+          for ( auto &coords : geomRing )
+          {
+            ring.append( QgsPointXY( coords.at( 0 ), coords.at( 1 ) ) );
+          }
+          polygon.append( ring );
+        }
+        geometry.append( polygon );
+      }
+      else if ( geom.at( "type" ) == "MultiPolygon" )
+      {
+        for ( auto &geomPart : geom.at( "coordinates" ) )
+        {
+          QgsPolygonXY part;
+          for ( auto &geomRing : geomPart )
+          {
+            QgsPolylineXY ring;
+            for ( auto &coords : geomRing )
+            {
+              ring.append( QgsPointXY( coords.at( 0 ), coords.at( 1 ) ) );
+            }
+            part.append( ring );
+          }
+          geometry.append( part );
+        }
+      }
+      else
+      {
+        QgsDebugMsg( QStringLiteral( "Unexpected geometry type: %1" ).arg( QString::fromStdString( geom.at( "type" ) ) ) );
+      }
+      si.geometry = QgsGeometry::fromMultiPolygonXY( geometry );
+    }
+    catch ( std::exception &e )
+    {
+      QgsDebugMsg( QStringLiteral( "Malformed geometry item: %1" ).arg( QString::fromStdString( e.what() ) ) );
+      return;
+    }
+
+    if ( si.uri.startsWith( QLatin1String( "./" ) ) )
+    {
+      // resolve relative path
+      si.uri = fInfo.absoluteDir().absoluteFilePath( si.uri );
+    }
+
+    if ( f["properties"].contains( "pc:schemas" ) )
+    {
+      for ( auto &schemaItem : f["properties"]["pc:schemas"] )
+      {
+        attributeNames.insert( QString::fromStdString( schemaItem["name"] ) );
+      }
+    }
+
+    if ( transform.isValid() && !transform.isShortCircuited() )
+      try
+      {
+        si.geometry.transform( transform );
+      }
+      catch ( QgsCsException & )
+      {
+        QgsDebugMsg( QStringLiteral( "Cannot transform geometry to layer crs." ) );
+        continue;
+      }
```

Comment:
would be good to have curly braces around this `if` - I feel a bit uneasy seeing such a multi-line statement without being wrapped in {} :slightly_smiling_face:

Reply:
Yikes! The try-catch curly braces tricked my eye